### ISSUE TITLE
[Feature] Allow HookAction to run only for specific hooks

### ISSUE THREAD

**Issue by shochdoerfer** [CaptainHookPhp/captainhook#71]
For most HookActions it's probably irrelevant to which Git hook they belong. For some HookActions like [captainhook-rejectpush](https://github.com/bitExpert/captainhook-rejectpush) a specific hook needs to be configured to make sure the HookAction works as intended.

I'd love to see a method in the HookAction that will be used by the Capt'n to check which hooks are valid for this action. When running the Capt'n each action configuration for the current Git hook needs to be checked and fail in case an ActionHook is used that "does not work" in the current Git hook configuration.

Basically I see two ways to accomplish this:
- Extend the existing `CaptainHook\App\Hook\Action` with a parent method that would allow the HookAction to be executed in all Git hooks (as it works now)
- Add a new marker interface with one specific method that HookActions need to implement. If the object instance does not implement the interface it would mean execute in all Git hooks (as it works now)

**Comment by sebastianfeldmann** [comment]
What do you thing of (implementation and naming)

```php
namespace CaptainHook\App\Hook;

interface Constrained
{
    /**
     * Returns the list of hooks where this action is applicable
     *
     * @return array<string>
     */
    public static function getApplicableHooks(): array;
}
```

Implementation would look something like this
```php

use CaptainHook\App\Hook\Action;
use CaptainHook\App\Hook\Constraint;

class MyAction implements Action, Constraint
{
    public static function getApplicableHooks(): array
    {
        return [
            Hooks::PRE_PUSH,
            Hooks::PRE_COMMIT
        ];
    }

    public function execute(Config $config, IO $io, Repository $repository, Config\Action $action): void
    {
        // something actionish
    }
}
```

**Comment by sebastianfeldmann** [comment]
BTW, this would work for conditions as well ;)

**Comment by heiglandreas** [comment]
I'd rather either use explicit ignore-Interfaces like an `IgnoresPushHook` interface that an action can implement (an explicit `HandlesPushHook` would be more intuitive but requires a lot more than an `instanceof`) or have the `getApplicableHooks` return a VO that has an `isPushAllowed` methid or something lime that. That way the returned value is more explicit than relying on a third party hopefully returning the right stuff in an array 😉

**Comment by sebastianfeldmann** [comment]
I'm not sure I like the idea of multiple interfaces because of convenience reasons.
Moving the implementation to the user side for the decision methods seems not a nice thing to do to the user that just wants to restrict his action to a specific hook.

But the value object idea is nice.
For user land that would mean something like this:

```php
public static function getHookRestriction(): Restriction
    {
        return new Restriction([
            Hooks::PRE_PUSH,
            Hooks::PRE_COMMIT
        ]);
    }
```

The internal implementation would change from
```php
return in_array(Hooks::PRE_COMMIT, $action->getApplicableHooks());
```
to 
```php
return $action->getHookRestriction()->isApplicableFor(Hooks::PRE_COMMIT);
```

**Comment by heiglandreas** [comment]
And if the user does something like 

```php
public static function getHookRestriction(): Restriction
    {
        return new Restriction([
            'Foo',
            1.5,
            'Some other invalid value',
        ]);
    }
```
the Constructor of the value-object can silently ignore the input or throw an `InvalidArgumentException`.

Though I'd personally would prefer

```php
final class Restriction {
    private $restrictions;
    private function __construct(string ...$restrictions)
    {
        foreach ($restrictions as $restriction) {
            if (! $this->isValidHook($hook)) {
                continue;
            }
            if (in_array($restriction, $this->restrictions)){
                continue;
            }
            $this->restrictions[] = $restriction;
        }
    }

    public static function fromArray(array $restrictions): self
    {
        return new self(...$restrictons);
    }

    public static function fromString(string $restriction): self
    {
        return new self($restriction);
    }

    public static function empty(): self
    {
        return new self([]);

    public function with(string $restriction): self
    {
        if (! $this->isValidHook($hook)) {
            return $this;
        }

        if (in_array($restriction, $this->restrictions)) {
            return $this;
        }

        $self = clone ($this);
        $self->restrictions[] = $restriction;

        return $self;
    }

    public function isApplicableFor(string $hook): bool
    {
        if (! $this->isValidHook($hook)) {
            return false;
        }

        return in_array($hook, $this->restrictions);
    }

    private function isValidHook(string $hook): bool
    {
        return in_array($hook, Hooks::AVAILABLE_HOOKS);
    }
}
```

So that we can then call 

```php
public static function getHookRestriction(): Restriction
    {
        return Restriction::fromArray([
            Hooks::PRE_PUSH,
            Hooks::PRE_COMMIT
        ]);
    }
```

or alternatively

```php
public static function getHookRestriction(): Restriction
{
    return Restriction::fromString(Hooks::PRE_PUSH)->with(Hooks::PRE_COMMIT);
}
```

or even

```php
public static function getHookRestriction(): Restriction
{
    return Restriction::empty()
        ->with(Hooks::PRE_PUSH)
        ->with(Hooks::PRE_COMMIT);
}
```

**Comment by shochdoerfer** [comment]
This escalated quickly :P 

I am fine with that one marker interface (name), just a slightly different approach: Why not make the Constrained interface extend the Action interface and name it ConstrainedAction?

The Restriction VO also seems nice. I like the VO API that @heiglandreas outlined above. 

Not sure though if it makes sense to apply this idea to the conditions, shouldn't they be generic? Technically we could make use of the conditions to restrict where the action can be run. But that would mean we pass the decision making to the user not the author of the action class.

**Comment by sebastianfeldmann** [comment]
I added an implementation to the current master
https://github.com/CaptainHookPhp/captainhook/blob/master/src/Hook/Constrained.php
https://github.com/CaptainHookPhp/captainhook/blob/master/src/Hook/Restriction.php

Currently I'm not extending the `Action` interface because I still think it makes sense to use the `Constrained` interface for `Conditions` as well.
Fo example, at the moment due of the way they are implemented the `Any` and `All` files changed conditions are only working on `post-merge` and `post-checkout` hooks.

**Comment by shochdoerfer** [comment]
WOW That was fast! LGTM.

**Comment by sebastianfeldmann** [comment]
I will put in into the third release candidate.

**Comment by sebastianfeldmann** [comment]
Tagged and released RC3 
https://github.com/CaptainHookPhp/captainhook/releases/tag/5.0.0-RC3

**Comment by sebastianfeldmann** [comment]
RC5 now support the `Constrained` interface for `Actions` and `Conditions` 
Alle built-in `Conditions` are implementing the interface if necessary.
